File: 2020/25/golang/breaker/combo-breaker.go
package breaker;

import (
	"fmt"
	"errors"
)

type ComboBreaker struct {
	MOD, MAX_LOOP_COUNT int
}

func (breaker ComboBreaker) GetLoopSize(subject int, pubKey int) (int, error) {
	generated_pub_key := subject
	for i := 1; i < breaker.MAX_LOOP_COUNT; i++ {
		generated_pub_key = (generated_pub_key * subject) % breaker.MOD
		if generated_pub_key == pubKey {
			return i + 1, nil
		}
	}
	return 0, errors.New(fmt.Sprintf("Could not find loop size. subject=%d, pub_key=%d", subject, pubKey))

}

func (breaker ComboBreaker) GetPrivateKey(loopSize int, pubKey int) int {
	private_key := 1

	for i := 0; i < loopSize; i++ {
		private_key = (private_key * pubKey) % breaker.MOD
	}
	
	return private_key
}

func (breaker ComboBreaker) Decrypt(subject int, cardPubKey int, doorPubKey int) (int, error) {

	var cardLoopSize, e = breaker.GetLoopSize(subject, cardPubKey)

	fmt.Println("Card loop size: ", cardLoopSize)

	if e != nil {
		return 0, e
	}

	privateKey := breaker.GetPrivateKey(cardLoopSize, doorPubKey)

	return privateKey, nil;
}

File: 2020/25/python/combo_breaker.py
class ComboBreaker:
    
    def __init__(self, max_loop_count = 20201227, mod = 20201227):
        self.__max_loop_count = max_loop_count
        self.__mod = mod

    def decrypt(self, subject, card_pub_key, door_pub_key):

        try:
            card_loop_size = self.__get_loop_size(subject, card_pub_key)

            print(f'Success to get loop size! card_loop_size={card_loop_size}', )
        
            private_key = self.__get_private_key(card_loop_size, door_pub_key)

            return private_key
        except Exception:
            print(f'Error! Check the parameters: subject={subject}, card_pub_key={card_pub_key}, door_pub_key={door_pub_key}')
            raise;

    def __get_private_key(self, loop_size, pub_key):
        private_key = 1

        for _ in range(loop_size):
            private_key = (private_key * pub_key) % self.__mod
        
        return private_key

    def __get_loop_size(self, subject, pub_key):
        generated_pub_key = subject

        for loopSize in range(1, self.__max_loop_count):
            generated_pub_key = (generated_pub_key * subject) % self.__mod
            if generated_pub_key == pub_key: 
                return loopSize + 1

        raise Exception(f'Could not find loop size. subject={subject}, pub_key={pub_key}')